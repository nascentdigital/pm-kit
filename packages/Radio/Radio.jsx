import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import { css } from '@emotion/core'
import generateId from '../../shared/utils/generateId/generateId.js'
import {
  parkGreen,
  softSandBrown,
  lilyGreen,
  lightTan,
  midnightGreen,
  forestFog,
  seafoam,
  birch,
} from '@pm-kit/colours'
import { motion } from 'framer-motion'

// Default Radio Button
const radio = css`
  opacity: 0;
  position: fixed;
  pointer-events: none;
  &:checked ~ label > div {
    background-color: initial;
  }
  &:checked ~ label > div > span {
    display: block;
  }
`
const basicRadioFake = css`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 17px;
  height: 17px;
  min-height: 17px;
  min-width: 17px;
  margin-top: -2px;
  outline: 0;
  line-height: 0;
  border-radius: 50%;
  border: 2px solid ${parkGreen};
  cursor: pointer;
  transition: border-color 0.1s linear, background-color 0.1s linear;
`

const radioFakeSelected = css`
  border: 2px solid ${softSandBrown};
`

const radioInner = css`
  display: none;
  width: 9px;
  height: 9px;
  border-radius: 50%;
  background-color: ${softSandBrown};
`

const basicLabel = css`
  display: flex;
  align-items: center;
  padding: 12px 18px;
  border-radius: 12px;
  cursor: pointer;
  overflow: hidden;
  background-position: right bottom;
  background-color: ${parkGreen};
  width: 100%;
`

const labelNotSelected = css`
  &:hover {
    background-color: ${lilyGreen};
  }
  background-color: ${lightTan};
`

const labelSelected = css`
  :hover {
    background-color: ${midnightGreen};
  }
  background-color: ${parkGreen};
  color: ${softSandBrown};
`

const labelText = css`
  margin-left: 1rem;
  font-size: 1em;
  font-weight: 700;
`

// Alternative Radio Button
const basicRadioFakeAlt = css`
  ${basicRadioFake}
  border: 2px solid ${parkGreen};
`

const basicLabelAlt = css`
  ${basicLabel}
  background-color: ${parkGreen};
`

const labelNotSelectedAlt = css`
  &:hover {
    background-color: ${birch};
  }
  background-color: ${forestFog};
`

const labelSelectedAlt = css`
  :hover {
    background-color: ${lilyGreen};
  }
  background-color: ${seafoam};
  color: ${parkGreen};
`
const radioFakeSelectedAlt = css`
  border: 2px solid ${parkGreen};
`
const radioInnerAlt = css`
  ${radioInner}
  background-color: ${parkGreen};
`

// Disabled Radio Button
const labelNotSelectedAltDisabled = css`
  ${labelNotSelectedAlt}
  opacity: 0.5;
  :hover {
    background-color: ${forestFog};
  }
  cursor: not-allowed;
`
const radioFakeSelectedAltDisabled = css`
  ${basicRadioFakeAlt}
  cursor: not-allowed;
`

export const Radio = ({
  checked,
  description,
  error,
  feedback,
  id,
  label,
  name,
  value,
  variant,
  type,
  forwardedRef,
  disabled,
  ...rest
}) => {
  const radioId = generateId(id, rest.name, label)
  const radioFakeStyle = [basicRadioFake]
  const labelStyle = [basicLabel, labelNotSelected]
  const radioFakeStyleAlt = [basicRadioFakeAlt]
  const labelStyleAlt = [basicLabelAlt, labelNotSelectedAlt]

  if (checked) {
    if (type === 'alternative') {
      labelStyleAlt.push(labelSelectedAlt)
      radioFakeStyleAlt.push(radioFakeSelectedAlt)
    } else {
      labelStyle.push(labelSelected)
      radioFakeStyle.push(radioFakeSelected)
    }
  } else if (disabled) {
    labelStyleAlt.push(labelNotSelectedAltDisabled)
    radioFakeStyleAlt.push(radioFakeSelectedAltDisabled)
  }

  const labelVariant = {
    whileTap: {
      scale: 0.96,
    },
  }

  return (
    <motion.div whileTap="whileTap">
      <input
        checked={checked}
        css={radio}
        data-testid="hidden-input"
        id={radioId.identity()}
        name={name}
        type="radio"
        value={value}
        disabled={disabled}
        {...rest}
      />
      <motion.label
        css={type === 'alternative' ? labelStyleAlt : labelStyle}
        data-testid="checkbox-label"
        htmlFor={radioId.identity()}
        variants={disabled ? '' : labelVariant}
        transition={disabled ? '' : { duration: 0.6 }}
      >
        <div css={type === 'alternative' ? radioFakeStyleAlt : radioFakeStyle} data-testid="fake-input">
          <span css={type === 'alternative' ? radioInnerAlt : radioInner} />
        </div>
        <span css={labelText}>{label}</span>
      </motion.label>
    </motion.div>
  )
}

const RadioWithRef = forwardRef((props, ref) => <Radio {...props} forwardedRef={ref} />)

RadioWithRef.propTypes = {
  /**
   * The label.
   */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Associate this radio with a group. Set as the HTML name attribute.
   */
  name: PropTypes.string.isRequired,
  /**
   * Control the appearance of the component
   */
  variant: PropTypes.oneOf(['borderless', 'bordered']),
  /**
   * Control the type of the component
   */
  type: PropTypes.oneOf(['default', 'alternative']),
  /**
   * The value. Must be unique within the group.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  /**
   * Use `checked` for controlled radio. For uncontrolled radio, use React's built-in `defaultChecked` prop.
   * See examples below for more details.
   */
  checked: PropTypes.bool,
  /**
   * Disabled state.
   */
  disabled: PropTypes.bool,
  /**
   * Description text below the radio.
   */
  description: PropTypes.string,
  /**
   * The id. Must be unique within the group. If no id is provided, one will be generated in the following format: `name_value`
   */
  id: PropTypes.string,
  /**
   * A feedback state.
   */
  feedback: PropTypes.oneOf(['error']),
  /**
   * An error message.
   */
  error: PropTypes.string,
}

RadioWithRef.defaultProps = {
  variant: 'borderless',
  type: 'default',
  disabled: false,
  description: undefined,
  feedback: undefined,
  error: undefined,
  checked: undefined,
  label: '',
  id: undefined,
}

Radio.displayName = 'Radio'

export default RadioWithRef
