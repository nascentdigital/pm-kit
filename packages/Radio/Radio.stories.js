import React, { useState } from 'react'
import Radio from './Radio'
import { css } from '@emotion/core'

// export default {
//   title: 'Design System|Radio',
//   component: Radio,
// }

export const Default = () => {
  const [clicked, setClicked] = useState('en')

  const handleSelectOption = (e) => {
    setClicked(e.target.value)
  }

  const radioButtonContainer = css`
    display: flex;
    flex-direction: column;

    > * :not(last-child) {
      margin-bottom: 10px;
    }
  `

  return (
    <div css={radioButtonContainer}>
      <Radio checked={clicked === 'en'} label="English" name="prefLang" onChange={handleSelectOption} value="en" />
      <Radio checked={clicked === 'fr'} label="French" name="prefLang" onChange={handleSelectOption} value="fr" />
    </div>
  )
}

export const Alternative = () => {
  const [clickedAlt, setClickedAlt] = useState('yes')

  const handleSelectOptionAlt = (e) => {
    setClickedAlt(e.target.value)
  }

  const radioButtonContainerAlt = css`
    display: flex;
    flex-direction: column;

    > * :not(last-child) {
      margin-bottom: 10px;
    }
  `

  return (
    <div css={radioButtonContainerAlt}>
      <Radio
        checked={clickedAlt === 'yes'}
        label="Yes"
        name="bool"
        onChange={handleSelectOptionAlt}
        value="yes"
        type="alternative"
      />
      <Radio
        checked={clickedAlt === 'no'}
        label="No"
        name="alternative"
        onChange={handleSelectOptionAlt}
        value="no"
        type="alternative"
      />
    </div>
  )
}

export const Disabled = () => {
  const [clickedAltDisabled, setClickedAltDisabled] = useState()

  const handleSelectOptionAltDisabled = (e) => {
    setClickedAltDisabled(e.target.value)
  }

  const radioButtonContainerAltDisabled = css`
    display: flex;
    flex-direction: column;

    > * :not(last-child) {
      margin-bottom: 10px;
    }
  `

  return (
    <div css={radioButtonContainerAltDisabled}>
      <Radio
        checked={clickedAltDisabled === 'disabled'}
        label="Disabled"
        name="disabled"
        onChange={handleSelectOptionAltDisabled}
        value="disabled"
        type="alternative"
        disabled
      />
    </div>
  )
}
