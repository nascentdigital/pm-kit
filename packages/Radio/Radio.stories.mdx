import { Meta, Story, Preview, Props, Subtitle } from '@storybook/addon-docs/blocks'
import Radio from './Radio.jsx'
import { Default } from './Radio.stories.js'
import { version } from './package.json'

<Meta title="Design System|Radio" componenent={Radio} />

# Radio
<Subtitle>
version {version}
</Subtitle>

## Usage
* To install `Radio` package with npm run terminal:

```jsx
npm i @pm-kit/radio

```
* Import `Radio` in your componenent to render

```jsx
import Radio from '@pm-kit/radio'

```

## Properties 

<Props of={Radio} />

## Stories

#### Default
<Preview>
<Story name="Default">
    <Default />
</Story>
</Preview>
