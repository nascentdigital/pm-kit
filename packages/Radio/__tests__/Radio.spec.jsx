import React from 'react'
import { shallow } from 'enzyme'

import Radio from '../Radio'

describe('Radio', () => {
  it('renders', () => {
    const radio = shallow(<Radio label="Testing" name="Test" value={true} />)

    expect(radio).toMatchSnapshot()
  })
})
