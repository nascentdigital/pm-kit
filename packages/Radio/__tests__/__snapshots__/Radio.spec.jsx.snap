// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders 1`] = `
<Radio
  disabled={false}
  forwardedRef={null}
  label="Testing"
  name="Test"
  type="default"
  value={true}
  variant="borderless"
/>
`;
